gen/wrap.jl

```diff
@@ -0,0 +1,187 @@
+]
+
+const replace_exprs = Dict{Expr,Expr}(
+    :(const VSIStatBuf = stat) => :(const VSIStatBuf = Cvoid),
+    :(const VSIStatBufL = __stat64) => :(const VSIStatBufL = Cvoid),
+)
+
+"""
+Custom rewriter for Clang.jl's C wrapper
+
+Gets called with all expressions in a header file, or all expressiong in a common file.
+If available, it adds docstrings before every expression, such that Clang.jl prints them
+on top of the expression. The expressions themselves get sent to `rewriter(::Expr)`` for
+further treatment.
+"""
+function rewriter(xs::Vector)
+    rewritten = Any[]
+    for x in xs
+        # Clang.jl inserts strings like "# Skipping MacroDefinition: X"
+        # keep these to get a sense of what we are missing
+        if x isa String
+            push!(rewritten, x)
+            continue
+        end
+        @assert x isa Expr
+
+        x in skip_exprs && continue
+        x = get(replace_exprs, x, x)
+
+        name = cname(x)
+        node = findnode(name, doc)
+        docstr = node === nothing ? "" : build_docstring(node)
+        isempty(docstr) || push!(rewritten, addquotes(docstr))
+        x2 = rewriter(x)
+        push!(rewritten, x2)
+    end
+    rewritten
+end
+
+"Rewrite expressions in the ways listed at the top of this file."
+function rewriter(x::Expr)
+    if @capture(x,
+        function f_(fargs__)
+            ccall(fname_, rettype_, argtypes_, argvalues__)
+        end
+    )
+        # it is a function wrapper around a ccall
+
+        # lowercase the function name
+        f′ = Symbol(lowercase(String(f)))
+
+        # bind the ccall such that we can easily wrap it
+        cc = :(ccall($fname, $rettype, $argtypes, $(argvalues...)))
+
+        # eval to check if it is Ptr{Cvoid} indirectly
+        # e.g. const GDALInfoOptions = Cvoid; Ptr{GDALInfoOptions}
+        # it works since we include common.jl, meaning we do need to run this twice
+        # in case a new run creates new const X = Cvoid
+        isptrvoid = try
+            eval(rettype) == Ptr{Cvoid}
+        catch  # UndefVarError
+            # it might still be true if common.jl is not yet complete, run twice to avoid this
+            false
+        end
+
+        cc′ = if rettype == :Cstring
+            :(unsafe_string($cc))
+        elseif rettype == :(Ptr{Cstring})
+            :(unsafe_loadstringlist($cc))
+        elseif isptrvoid
+            :(failsafe($cc))
+        else
+            cc
+        end
+
+        # stitch the modified function expression back together
+        :(function $f′($(fargs...))
+            $cc′
+        end) |> prettify
+    else
+        # do not modify expressions that are no ccall function wrappers
+        x
+    end
+end
+
+# parse GDAL's Doxygen XML file
+const doc = readxml(xmlpath)
+
+# custom function that prevents overwriting GDAL.jl by mapping gdal.h to gdal_h.jl
+function header_outputfile(h)
+    stem = splitext(basename(h))[1]
+    stem = stem == "gdal" ? "gdal_h" : stem
```

Comment:
nit: I think it might be clearer to write it as
```julia
if splitext(basename(h))[1] == "gdal"
    return joinpath(output_dir, "gdal_h.jl")
else
    return joinpath(output_dir, "gdal.jl")
end
```

Reply:
Yes that's not quite correct but I get what you're saying ;)